### PR TITLE
Multiple fdata per record

### PR DESCRIPTION
**Pull request by ErlendHaa**

### CODE REVIEW

File: python/dlisio/ext/core.cpp [hunk 1]
```diff
@@ -342,6 +383,11 @@ noexcept (false) {
 
         /* get frame number and slots */
         while (ptr < end) {
+            if (frames == allocated_rows) {
+                resize(frames * 2);
+                dst += (frames * itemsize);
```
[the remark this answers is not in the document]
Comment (jokva) [reply python/dlisio/ext/core.cpp]: I made it be outside for that specific reason - so all advancing of dst happens inside the loop.

The entire thing is to be refactored next anyway, as the function is somewhat unwieldy.

File: python/dlisio/ext/core.cpp [hunk 2]
```diff
@@ -295,12 +295,14 @@ std::string fingerprint(const std::string& type,
...
-                const std::vector< int >& indices,
-                py::object dstobj)
+py::object read_fdata(const char* pre_fmt,
+                      const char* fmt,
+                      const char* post_fmt,
+                      dl::stream& file,
+                      const std::vector< int >& indices,
+                      std::size_t itemsize,
+                      py::object alloc,
+                      const std::vector< int >& object_offsets)
```
[the remark this answers is not in the document]
Comment (jokva) [reply python/dlisio/ext/core.cpp]: Good point, it's a remnant from an earlier design, but it is now managed by numpy instead. I'll remove it.

File: python/dlisio/ext/core.cpp [hunk 3]
```diff
@@ -309,9 +311,47 @@ noexcept (false) {
...
+     * and then carefully restored to the new memory.
+     */
+    auto resize = [&](std::size_t n) {
+        info = py::buffer_info {};
+        dstb = py::buffer {};
+        dstobj.attr("resize")(n);
+        allocated_rows = n;
+        dstb = py::buffer(dstobj);
+        info = dstb.request(true);
+        dst = static_cast< unsigned char* >(info.ptr);
```
[the remark this answers is not in the document]
Comment (jokva) [reply python/dlisio/ext/core.cpp]: Codacy gets it wrong.

File: python/dlisio/plumbing/frame.py [hunk 4]
```diff
@@ -264,24 +265,36 @@ def fmtstr(self):
...
+            curves with dtype = self.dtype
+
+        See also
         --------
+        Channel.curves : Access the curve-data directly through the Channel
+          objects
+        Frame.dtype : dtype of the array
 
+        Examples
+        --------
```
[the remark this answers is not in the document]
Comment (jokva) [reply python/dlisio/plumbing/frame.py]: It is.

File: python/dlisio/plumbing/frame.py [hunk 5]
```diff
@@ -154,17 +154,17 @@ def create(cls, obj, name = None, type = None, file = None):
...
-        and copynumber.
+        If all curve mnemonics are unique, then dtype.names == ['FRAMENO'] +
+        [ch.name for ch in self.channels]. If there are more than one channel
+        with the same name for this frame, all duplicated mnemonics are
+        enriched with origin and copynumber.
 
         Consider a frame with the channels mnemonics [('TIME', 0, 0), ('TDEP',
         0, 0), ('TIME, 1, 0)]. The dtype names for this frame would be
-        ('TIME.0.0', 'TDEP', 'TIME.1.0').
+        ('FRAMENO', 'TIME.0.0', 'TDEP', 'TIME.1.0').
```
[the remark this answers is not in the document]
Comment (jokva) [reply python/dlisio/plumbing/frame.py]: Yes, and the index property should implement the rules where it falls back to frameno.

File: python/dlisio/ext/core.cpp [hunk 6]
```diff
@@ -333,14 +342,6 @@ noexcept (false) {
...
-            ptr = dlis_uvari(ptr, &frameno);
-
-            if (frameno != expected_frameno) {
-                // TODO: warning
-                // const auto msg = 'Non-sequential frames. expected = {}, current = {}'
-            }
-
             auto assert_overflow = [end](const char* ptr, int skip) {
                 if (ptr + skip > end) {
                     const auto msg = "corrupted record: fmtstr would read past end";
```
[the remark this answers is not in the document]
Comment (jokva) [reply python/dlisio/ext/core.cpp]: Yep, and I hope better diagnostics will be supported when its refactored.

**Comment by ErlendHaa** [comment]
closes #165
closes #166

**Pull request merged**